### PR TITLE
Middleware to validate request for d.velop lifecylce events from d.velop cloud center

### PR DESCRIPTION
**Pull request by dvchsom**
You can use this middleware to validate d.velop lifecylce events from d.velop cloud center. 
Also you can use a validation method it self, when you will not use a middleware for lifecylce event validation.

### CODE REVIEW

File: requestsignature/requestsignature.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+	"strings"
+	"time"
+)
+
+type RequestSigner interface {
+	ValidateSignedRequest(req *http.Request) error
+}
+
+type RequestSignatureDto struct {
+					EventType string `json:"type"`
```
Comment (weberaj) [inline requestsignature/requestsignature.go:25]: Clean up indentation.

File: requestsignature/requestsignature.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+//			if err != nil {
+//				log.Print(err)
+//				http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+//				return
+//			}
+//			doSomeStuff(eventDto)
+//		})
+//	}
+
+func HandleSignMiddleware(appSecret []byte, timeNow func() time.Time) func(http.Handler) http.Handler {
```
Comment (weberaj) [inline requestsignature/requestsignature.go:80]: Does the timeNow function have to be public or can it be an implemenation detail. For testability you have to change your package scope to be an internal test of the package.

File: requestsignature/requestsignature.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+
+			if appSecret == nil {
+				log.Print("error validation signed request because app secret has not been configured")
```
Comment (weberaj) [inline requestsignature/requestsignature.go:85]: Make the logger changable for the consumer of this package.

File: requestsignature/requestsignature.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+func NewRequestSigner(appSecret []byte, timeNow func() time.Time) RequestSigner {
+	return &requestSigner{
+		appSecret,
+		timeNow,
+	}
+}
+
+// validate signed request as function
+func (signer *requestSigner) ValidateSignedRequest(req *http.Request) error {
+	if signer.appSecret == nil {
```
Comment (weberaj) [inline requestsignature/requestsignature.go:181]: Use len(signer.appSecret) == 0

File: requestsignature/requestsignature.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+	err = signer.isTimestampValid(req)
+	if err != nil {
+		return err
+	}
+	normalizedRequestHash, err := signer.getHexHashForNormalizedHeaders(req)
+	if err != nil {
+		return err
+	}
+	log.Print("normalized request hex hash: ", normalizedRequestHash)
```
Comment (weberaj) [inline requestsignature/requestsignature.go:202]: Remove this log statement.

File: requestsignature/requestsignature.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+	hmacHexValue := signer.getHmacHash(normalizedRequestHash)
+	log.Print("hmac hex hash: ", hmacHexValue)
```
Comment (weberaj) [inline requestsignature/requestsignature.go:205]: Remove this log statement.

File: requestsignature/requestsignature.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+	normalizedRequest = append(normalizedRequest, req.URL.RawQuery)
+	normalizedRequest = append(normalizedRequest, fmt.Sprintf("%v\n", strings.Join(normalizedHeaders, "\n")))
+	normalizedRequest = append(normalizedRequest, signer.getHexHashedPayload(body))
+
+	return normalizedRequest
+}
+
+func (signer *requestSigner) getBodyFromRequest(req *http.Request) ([]byte, error) {
+	if req.GetBody != nil {
+		log.Print("get copy of request body")
```
Comment (weberaj) [inline requestsignature/requestsignature.go:291]: Remove this log statement.

File: requestsignature/requestsignature.go [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+			return nil,  err
+		}
+
+		body, err := ioutil.ReadAll(bodyReader)
+		if err != nil {
+			return nil, err
+		}
+
+		return  body,nil
+	} else {
```
Comment (weberaj) [inline requestsignature/requestsignature.go:305]: Remove redundant else statement.

File: requestsignature/requestsignature.go [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+		log.Print("request.GetBody is nil. Read body and create new request body (reader) with copy of body")
```
Comment (weberaj) [inline requestsignature/requestsignature.go:306]: Remove this log statement.

File: requestsignature/requestsignature.go [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+		if err != nil {
+			return nil,err
+		}
+
+		return body, nil
+	}
+}
+
+func (signer *requestSigner) getHexHashedPayload(payload []byte) string {
+	log.Print("payload: ", string(payload))
```
Comment (weberaj) [inline requestsignature/requestsignature.go:318]: Remove this log statement.

File: requestsignature/requestsignature_test.go [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+)
+
+const timestampHeader = "x-dv-signature-timestamp"
+const algorithmHeader = "x-dv-signature-algorithm"
+const signedHeadersHeader = "x-dv-signature-headers"
+const authorizationHeader = "authorization"
+
+const algorithm = "DV1-HMAC-SHA256"
+
+func TestRequestSigner_ValidateSignedRequest_HappyPath_Working(t *testing.T) {
```
Comment (weberaj) [inline requestsignature/requestsignature_test.go:21]: Test cases could be handled more elegant with table driven tests.

File: requestsignature/requestsignature.go [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+	if req.GetBody != nil {
+		log.Print("get copy of request body")
+
+		var bodyReader io.Reader
+		bodyReader, err := req.GetBody()
+		if err != nil {
+			return nil,  err
+		}
+
+		body, err := ioutil.ReadAll(bodyReader)
```
Comment (thomasmkt) [inline requestsignature/requestsignature.go:299]: At this point you read the body of the request. The body will not be visible for other HTTP-Handler. You can simply add the read body as a NopCloser back to the Request. 

req.Body = ioutil.NopCloser(bytes.NewReader(body))

File: requestsignature/requestsignature.go [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+
+		return  body,nil
+	} else {
+		log.Print("request.GetBody is nil. Read body and create new request body (reader) with copy of body")
+
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return nil,err
+		}
+
```
Comment (thomasmkt) [inline requestsignature/requestsignature.go:312]: At this point you read the body of the request. The body will not be visible for other HTTP-Handler. You can simply add the read body as a NopCloser back to the Request. 

req.Body = ioutil.NopCloser(bytes.NewReader(body))

**Pull request closed without merging**